_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+  automation devices, etc.
+- The above types of embedded systems are primarily constrained by memory and
+  code space limitations, so the information below takes this into account.
+- The examples below are implemented for the ARM Cortex-M series of MCU's. Many
+  of the concepts can be generalized to other architectures.
+
+## Making the Most of Asserts
+
+To get the most value from asserting, it is important to set up some postmortem
+infrastructure. When an assert takes place (or even a system fault), there is _a
+lot_ of context and data one can capture to help debug these issues without a
+debugger attached to the device. These include but are not limited to:
+
+- the file and line number the assert took place,
+- the registers at the time of assert
+- values or function pointers on the stack, to help reassemble the backtrace
+- the logs that took place before the assert and possibly after the reboot took
+  place
+- for those using MyNewt[^1], ESP-IDF[^2], or have Memfault[^3] integrated, a
+  coredump can be captured at the time of the assert, which will provide a full
+  dump of system memory and registers.
+
+With all of the above, especially coredumps, it becomes advantageous to add as
+many asserts to the system as possible, especially for capturing those 1 in
+10,000 hour bugs. This allows developers to quickly track down and investigate
+bugs affecting users in the field, all without having to reproduce the issue on
+a development board.
+
+## Common Usages of Asserts
+
+Every firmware has places where it is and isn't appropriate to assert. The
+examples below likely apply to most non-safety-critical systems.
+
+### When malloc shouldn't fail
+
+Dynamic memory allocation is common in embedded systems, especially in today's
+Internet connected consumer electronics. These devices need to be able to
+receive an influx Bluetooth packets, use large HTTP buffers, and possibly load
+assets or fonts into RAM for a user interface. Keeping around statically
+allocated buffers for all of these functions would be irresponsible, especially
+for the devices that measure RAM in kilobytes. In these devices, dynamic memory
+allocation is a must.
+
+There are also places where malloc could fail, and where **it shouldn't**.
+
+For example, let's give an example of dynamically allocated mutexes.
+
+```c
+Mutex *mutex_create(void) {
+  Mutex *mutex = malloc(sizeof(Mutex));
+  ASSERT(mutex);
+  mutex_init(mutex);
+
+  return mutex;
+}
+```
+
+This malloc **should never fail**, because it would mean the system is out of
+RAM and should be rebooted to recover itself.
+
+### Invalid arguments passed into function
+
+Argument checking is one of the best (and easiest) places to assert.
+
+We can check for NULL pointers
+
+```c
+void handle_event(Event *evt) {
+  ASSERT(evt);
+  ...
+}
+```
+
+or check sizes of buffers
+
+```c
+void read_packet(void *buf, size_t buf_len) {
+  ASSERT(buf_len >= sizeof(Packet));
+  memcpy(...)
+}
+
+```
+
```

Comment:
Add a sentence or two on why asserting is better than painstakingly checking arguments. Oftentimes, passing a NULL pointer is a bug, not something that should be recovered from gracefully.